### PR TITLE
Buffering of custom events and fixed tests

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 1]
```diff
@@ -94,14 +96,27 @@ class InstrumentationService {
...
     this.scrollHandlerRegistered_ = false;
 
     /** @private {!Observable<Event>} */
     this.scrollObservable_ = new Observable();
+
+    /** @private {!Object<string, !Observable<!AnalyticsEvent>>} */
+    this.customEventObservers_ = {};
+
+    /**
+     * Some early events have to be buffered because there's no way to predict
```
Comment (erwinmombay) [inline extensions/amp-analytics/0.1/instrumentation.js]: can we remove "some" here, makes me think that only certain types of events can be buffered

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 2]
```diff
@@ -152,12 +167,25 @@ class InstrumentationService {
...
       if (!observers) {
         observers = new Observable();
-        this.observers_[eventType] = observers;
+        this.customEventObservers_[eventType] = observers;
       }
       observers.add(listener);
+
+      // Push recent events if any.
+      if (this.customEventBuffer_) {
+        this.customEventBuffer_.forEach(event => {
```
Comment (erwinmombay) [inline extensions/amp-analytics/0.1/instrumentation.js]: /cc @jridgewell 

should this be flushed asynchronously?

Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: Yes. Let's do.

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 3]
```diff
@@ -94,14 +96,27 @@ class InstrumentationService {
...
+    /**
+     * Some early events have to be buffered because there's no way to predict
+     * how fast all `amp-analytics` elements will be instrumented.
+     * @private {!Array<!AnalyticsEvent>|undefined}
+     */
+    this.customEventBuffer_ = [];
+
+    // Stop buffering of custom events after 10 seconds. Assumption is that all
+    // `amp-analytics` elements will have been instrumented by this time.
+    timer.delay(() => {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/instrumentation.js]: Could we do a poll check instead?

Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: There's unfortunately nothing to check for - there are no conditions that would tell us that a new client should be pending.

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 4]
```diff
@@ -94,14 +96,27 @@ class InstrumentationService {
...
+
+    /** @private {!Object<string, !Observable<!AnalyticsEvent>>} */
+    this.customEventObservers_ = {};
+
+    /**
+     * Some early events have to be buffered because there's no way to predict
+     * how fast all `amp-analytics` elements will be instrumented.
+     * @private {!Array<!AnalyticsEvent>|undefined}
+     */
+    this.customEventBuffer_ = [];
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/instrumentation.js]: Should we use a map like `#customEventObservers_` so we don't have to loop through _all_ events every time X event listener is added?

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: It'd have to be a map of list. We could do. But in reality, very very few custom events are possible in this case. If you feel strongly, I'll switch.

Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: Actually, i'll just go ahead and do it to avoid future regression.

Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

**Comment by dvoytenko** [comment]
@erwinmombay PTAL.

**Comment by amphtml-team** [comment]
Lgtm!

On Fri, Feb 12, 2016, 5:28 PM Dima Voytenko <notifications@github.com>
wrote:

> @erwinmombay <https://github.com/erwinmombay> PTAL.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ampproject/amphtml/pull/1998#issuecomment-183553261>.
>
> --
> You received this message because you are subscribed to the Google Groups
> "amphtml-eng-github" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to amphtml-eng-github+unsubscribe@google.com.
> To post to this group, send email to amphtml-eng-github@google.com.
> To view this discussion on the web visit
> https://groups.google.com/a/google.com/d/msgid/amphtml-eng-github/ampproject/amphtml/pull/1998/c183553261%40github.com
> <https://groups.google.com/a/google.com/d/msgid/amphtml-eng-github/ampproject/amphtml/pull/1998/c183553261%40github.com?utm_medium=email&utm_source=footer>
> .
>
> --
> You received this message because you are subscribed to the Google Groups
> "amphtml-eng" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to amphtml-eng+unsubscribe@google.com.
> To post to this group, send email to amphtml-eng@google.com.
> To view this discussion on the web visit
> https://groups.google.com/a/google.com/d/msgid/amphtml-eng/ampproject/amphtml/pull/1998/c183553261%40github.com
> <https://groups.google.com/a/google.com/d/msgid/amphtml-eng/ampproject/amphtml/pull/1998/c183553261%40github.com?utm_medium=email&utm_source=footer>
> .
>
-- 

Erwin Mombay | UX Engineer | erwinm@google.com | 650-772-9166

**Pull request merged**